Concept/nav 116 range raptor implementation

clukas1 opened the pull request:
First throw at range raptor. Think it would be interesting to benchmark against a reverse time direction routing request. I.e. for all pareto optimal simple raptor departure requests run reverse arrival requests, this should return the same results as the range raptor and may or may not be faster..

munterfi reviewed (changes requested):
Thanks, @clukas1!

Nice work, also getting quite complex. I have some points concerning documentation and configuration of the range:

- I tried to follow what happens in the range raptor, but honestly I am a bit confused with all the different types of offsets. Could please document this a bit more in detail.

- From my point of view the range is something query specific, since the user could set the range he likes to depart or arrive. Default I would set it to -1 in the QueryConfig of the service, and allow the UI to configure a sensible default value (e.g. 30minutes). What is the reason for having it in the application config?

- If we still keep it in the application config, I would introduce a Strategy pattern, since the check:

  ```java
        if (raptorRange <= 0) {
            doRounds(markedStops);
        } else {
            doRangeRaptor(markedStops);
        }
  ```
  could be avoided since it has always the same result for the configured raptor and the corresponding strategy (rangeQuery or simpleQuery) could be set on the raptor router via the constructor.

- If we move it to the user specific query config of the service (in this case then not the raptor query config), we could also consider extending the `RaptorAlgorithm` interface with the methods: `earliestArrivalRangeQuery` or `latestDepartureRangeQuery`. Reason behind this idea is that in the paper those are distinct versions of the raptor algorithm and in perspective also a `multicriteriaQuery` method could be added to the interface. We have the added versions of the arrival and departure cases.

Other from this the concept is good to merge from my side.

Reviewing the code now for some hours in the last days, I feel we should also again focus on breaking the Raptor router down into smaller, easier to understand units in the current sprint, otherwise we will have a hard time maintaining it in future and also reviewing is quite a time-consuming pain 😄

clukas1 commented on src/main/java/ch/naviqore/raptor/router/RouteScanner.java:
```diff
@@ -388,7 +389,128 @@ private boolean checkIfTripIsPossibleAndUpdateMarks(int targetTime, ActiveTrip a
+     * @return the trip offsets for the stop in the given range.
+     */
+    ArrayList<Integer> getTripOffsetsForStop(int stopIdx, int range) {
+        ArrayList<Integer> tripOffsets = new ArrayList<>();
```
c7eb452

clukas1 commented on src/main/java/ch/naviqore/raptor/router/RouteScanner.java:
```diff
@@ -388,7 +389,128 @@ private boolean checkIfTripIsPossibleAndUpdateMarks(int targetTime, ActiveTrip a
+     * @param range   the range to get the trip offsets for.
+     * @return the trip offsets for the stop in the given range.
+     */
+    ArrayList<Integer> getTripOffsetsForStop(int stopIdx, int range) {
```
c7eb452

clukas1 commented on src/main/java/ch/naviqore/raptor/router/Query.java:
```diff
@@ -117,8 +167,22 @@ List<StopLabelsAndTimes.Label[]> run() {
 
-        return stopLabelsAndTimes.getBestLabelsPerRound();
+    private List<Integer> getRangeOffsets(Set<Integer> markedStops, RouteScanner routeScanner) {
+        ArrayList<Integer> rangeOffsets = new ArrayList<>();
```
c7eb452

clukas1 commented on src/main/java/ch/naviqore/raptor/router/Query.java:
```diff
@@ -117,8 +167,22 @@ List<StopLabelsAndTimes.Label[]> run() {
+    private List<Integer> getRangeOffsets(Set<Integer> markedStops, RouteScanner routeScanner) {
+        ArrayList<Integer> rangeOffsets = new ArrayList<>();
+        for (int stopIdx : markedStops) {
+            ArrayList<Integer> stopRangeOffsets = routeScanner.getTripOffsetsForStop(stopIdx, raptorRange);
```
c7eb452

clukas1 commented on src/main/java/ch/naviqore/raptor/router/Query.java:
```diff
@@ -117,8 +167,22 @@ List<StopLabelsAndTimes.Label[]> run() {
+    }
 
-        return stopLabelsAndTimes.getBestLabelsPerRound();
+    private List<Integer> getRangeOffsets(Set<Integer> markedStops, RouteScanner routeScanner) {
```
821d782

clukas1 commented on src/main/java/ch/naviqore/raptor/router/Query.java:
```diff
@@ -117,8 +167,22 @@ List<StopLabelsAndTimes.Label[]> run() {
+        for (int stopIdx : markedStops) {
+            ArrayList<Integer> stopRangeOffsets = routeScanner.getTripOffsetsForStop(stopIdx, raptorRange);
+            for (int i = 0; i < stopRangeOffsets.size(); i++) {
+                if (rangeOffsets.size() == i) {
```
a07a3ce

clukas1 commented on src/main/java/ch/naviqore/raptor/router/RouteScanner.java:
```diff
@@ -388,7 +389,128 @@ private boolean checkIfTripIsPossibleAndUpdateMarks(int targetTime, ActiveTrip a
+     *
+     * @param stopIdx the index of the stop to get the trip offsets for.
+     * @param range   the range to get the trip offsets for.
+     * @return the trip offsets for the stop in the given range.
```
See Docstring further above:

> The trip offsets are expressed in a positive number of seconds from the reference time (in case of arrival scanning these values should be subtracted from the latest arrival time).

clukas1 commented on src/main/java/ch/naviqore/raptor/router/Query.java:
```diff
@@ -90,17 +91,66 @@ class Query {
+            int timeFactor = timeType == TimeType.DEPARTURE ? 1 : -1;
+            log.debug("Running rounds with range offset {}", rangeOffset);
+
+            // set source times to the source times of the previous round
```
We need to reset them because the targetTime of the initial label has to be updated for each range offset.
So if our range raptor spawns for minutes 0, 15, 30, we want to set the Initial Label Target time to departure time + 30 minutes, then departure time + 15 minutes and finally departure time.

clukas1 commented:
@munterfi: Thanks for your review, I hope I've addressed your comments well enough. Explaining all the steps in the code is relatively difficult.

Regarding your comment about the QueryConfig. I've spent some time thinking about it and don't think the current or the proposed solution are really ideal.

The reason I wouldn't include it into the QueryConfig is that it is very technical and something a regular user would never set in an user interface. I see some arguments for adding it to the QueryConfig of the raptor package, as the service could dynamically decide which range is appropriate for a request based on distance between start and end stop. Say 15 Minutes for distances < 10 km, 30 min for < 100 km and 1 h for the rest. But this could theoretically also be done by the service in the current implementation using the RaptorConfig reference and updating the raptorRange there before routing.

The main problem is, the problem of the range config value is that is extremely specific to our raptor implementation. So it's not generally applicable for other raptor implementations according to our interface (a problem which also holds true if we ever would implement a McRaptor). And also problematic since the `PublicTransitService` interface in the service is generally not restricted to only raptor implementations.

> If we still keep it in the application config, I would introduce a Strategy pattern, since the check:

      if (raptorRange <= 0) {
          doRounds(markedStops);
      } else {
          doRangeRaptor(markedStops);
      }

> could be avoided since it has always the same result for the configured raptor and the corresponding strategy (rangeQuery or simpleQuery) could be set on the raptor router via the constructor.

I honestly don't see a meaningful and easy way to do this.

> If we move it to the user specific query config of the service (in this case then not the raptor query config), we could also consider extending the RaptorAlgorithm interface with the methods: earliestArrivalRangeQuery and latestDepartureRangeQuery. Reason behind this idea is that in the paper those are distinct versions of the raptor algorithm and in perspective also a multicriteriaQuery method could be added to the interface. We have the added versions of the arrival and departure cases.

In this case, I'm against extending the interface. I would rather create three raptor implementations (Simple, Range, and MultiCriteria) which all implement the earliest, latest, and isoLine method. And have algorithm specific query configs handled before calling the routing methods. e.g. setRange(x) for range raptor or .setCriteria(y) for multicriteria. Although having a simple implementation probably does not make sense as it would be the range raptor with range 0. Which actually brings us to our current state.

munterfi reviewed (approved):
Thanks for the adjustments, looks good to merge.
